source/cellular_bg770.c

```diff
@@ -407,167 +666,324 @@ CellularError_t Cellular_ModuleEnableUE( CellularContext_t * pContext )
             }
             else
             {
-                LogError( ( "Cellular_ModuleEnableUE: '%s' command failed.", atReqGetNoResult.pAtCmd ) );
+                LogError( ( "Cellular_ModuleEnableUE: '%s' command failed (err: %s [%d]).",
+                            atReqGetNoResult.pAtCmd, getCellularErrorString(cellularStatus), cellularStatus ) );
             }
         }
 
-        #ifndef CELLULAR_CONFIG_DISABLE_FLOW_CONTROL
-            if( cellularStatus == CELLULAR_SUCCESS )
+#ifndef CELLULAR_CONFIG_DISABLE_FLOW_CONTROL
+        if( cellularStatus == CELLULAR_SUCCESS )
+        {
+            vTaskDelay( SHORT_DELAY_ticks );
+
+            static const BG770FlowControlState_t desiredFlowControlState =
             {
-                vTaskDelay( SHORT_DELAY_ticks );
+                .dceByDTE = BG770_FLOW_CONTROL_TYPE_HARDWARE,
+                .dteByDCE = BG770_FLOW_CONTROL_TYPE_HARDWARE
+            };
+            BG770FlowControlState_t flowControlState =
+            {
+                .dceByDTE = BG770_FLOW_CONTROL_TYPE_UNKNOWN,
+                .dteByDCE = BG770_FLOW_CONTROL_TYPE_UNKNOWN
+            };
+            cellularStatus = _GetFlowControlStateWithRetryTimeout(
+                    pContext, &flowControlState, ENABLE_MODULE_UE_RETRY_TIMEOUT_MS,
+                    ENABLE_MODULE_UE_RETRY_EXP_BACKOFF_INTER_COMMAND_BASE_MS );
+            if( cellularStatus != CELLULAR_SUCCESS ||
+                desiredFlowControlState.dceByDTE != flowControlState.dceByDTE ||
+                desiredFlowControlState.dteByDCE != flowControlState.dteByDCE )
+            {
+                if( cellularStatus != CELLULAR_SUCCESS )
+                {
+                    LogError( ( "Cellular_ModuleEnableUE: Could not get hardware flow control state, assuming not already set." ) );
+                }
 
-                /* Enable RTS/CTS hardware flow control. */
-                atReqGetNoResult.pAtCmd = "AT+IFC=2,2";
-                cellularStatus = sendAtCommandWithRetryTimeout( pContext, &atReqGetNoResult );
+                /**< No retry, even if miss 'OK' the flow control may have changed */
+                cellularStatus = _SetFlowControlState( pContext, desiredFlowControlState );
                 if( cellularStatus == CELLULAR_SUCCESS )
                 {
-                    LogInfo( ( "Cellular_ModuleEnableUE: '%s' command success.", atReqGetNoResult.pAtCmd ) );
+                    LogInfo( ( "Cellular_ModuleEnableUE: Set hardware flow control command success." ) );
+                    if( configSkipPostHWFlowControlSetupIfChanged )
+                    {
+                        LogInfo( ( "Cellular_ModuleEnableUE: Full initialization skipped based on flow control state change." ) );
+                        fullInitSkippedResult = CELLULAR_FULL_INIT_SKIPPED_RESULT_YES;

```

Comment:
The point is so that after Cellular_Init() returns, the code can query whether the init terminated early due to hardware flow control change. The idea is to only query this when Cellular_Init() isn't running therefore no chance of race condition.